plugin/src/main/java/org/opensearch/ml/action/undeploy/TransportUndeployModelAction.java

```diff
@@ -96,68 +94,90 @@ protected MLUndeployModelNodesResponse newResponse(
                         String modelId = entry.getKey();
-                        if (!removedNodeMap.containsKey(modelId)) {
-                            removedNodeMap.put(modelId, new ArrayList<>());
+                        if (!actualRemovedNodesMap.containsKey(modelId)) {
+                            actualRemovedNodesMap.put(modelId, new ArrayList<>());
                         }
-                        removedNodeMap.get(modelId).add(r.getNode().getId());
-                    }
-                    if (NOT_FOUND.equals(status)) {
-                        notFoundModels.add(entry.getKey());
+                        actualRemovedNodesMap.get(modelId).add(r.getNode().getId());
                     }
                 }
-                notFoundModels.forEach(m -> modelUndeployStatus.remove(m));
             });
-            Map<String, String[]> removedNodes = new HashMap<>();
-            for (Map.Entry<String, List<String>> entry : removedNodeMap.entrySet()) {
-                removedNodes.put(entry.getKey(), entry.getValue().toArray(new String[0]));
-                log.debug("removed node for model: {}, {}", entry.getKey(), Arrays.toString(entry.getValue().toArray(new String[0])));
-            }
-            MLSyncUpInput syncUpInput = MLSyncUpInput.builder().removedWorkerNodes(removedNodes).build();
+
+            MLSyncUpInput syncUpInput = MLSyncUpInput
+                .builder()
+                .removedWorkerNodes(covertRemoveNodesMapForSyncUp(actualRemovedNodesMap))
+                .build();
 
             MLSyncUpNodesRequest syncUpRequest = new MLSyncUpNodesRequest(nodeFilter.getAllNodes(), syncUpInput);
             try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
-                if (removedNodeMap.size() > 0) {
+                if (actualRemovedNodesMap.size() > 0) {
                     BulkRequest bulkRequest = new BulkRequest();
-                    for (String modelId : removedNodeMap.keySet()) {
+                    Map<String, Boolean> deployToAllNodes = new HashMap<>();
+                    for (String modelId : actualRemovedNodesMap.keySet()) {
                         UpdateRequest updateRequest = new UpdateRequest();
-                        int removedNodeCount = removedNodeMap.get(modelId).size();
-                        MLModelState mlModelState = modelWorkNodeCounts.get(modelId) > removedNodeCount
-                            ? MLModelState.PARTIALLY_DEPLOYED
+                        List<String> removedNodes = actualRemovedNodesMap.get(modelId);
+                        int removedNodeCount = removedNodes.size();
+                        // User remove partial nodes but not all, we regard this is a new user deploy plan, so the model status is deployed.
+                        MLModelState mlModelState = modelWorkNodesBeforeRemoval.get(modelId).length > removedNodeCount
+                            ? MLModelState.DEPLOYED
```

Comment:
What if the original state is PARTIALLY_DEPLOYED?

Reply:
Changed the checking logic and added TODO, we need to fix some corner case in future versions.